pkg/volume/util/util.go

```diff
@@ -500,28 +500,85 @@ func MakeAbsolutePath(goos, path string) string {
 func MapBlockVolume(
+	blkUtil volumepathhandler.BlockVolumePathHandler,
 	devicePath,
 	globalMapPath,
 	podVolumeMapPath,
 	volumeMapName string,
 	podUID utypes.UID,
 ) error {
-	blkUtil := volumepathhandler.NewBlockVolumePathHandler()
-
-	// map devicePath to global node path
-	mapErr := blkUtil.MapDevice(devicePath, globalMapPath, string(podUID))
+	// map devicePath to global node path as bind mount
+	mapErr := blkUtil.MapDevice(devicePath, globalMapPath, string(podUID), true /* bindMount */)
 	if mapErr != nil {
-		return mapErr
+		return fmt.Errorf("blkUtil.MapDevice failed. devicePath: %s, globalMapPath:%s, podUID: %s, bindMount: %v: %v",
+			devicePath, globalMapPath, string(podUID), true, mapErr)
 	}
 
 	// map devicePath to pod volume path
-	mapErr = blkUtil.MapDevice(devicePath, podVolumeMapPath, volumeMapName)
+	mapErr = blkUtil.MapDevice(devicePath, podVolumeMapPath, volumeMapName, false /* bindMount */)
+	if mapErr != nil {
+		return fmt.Errorf("blkUtil.MapDevice failed. devicePath: %s, podVolumeMapPath:%s, volumeMapName: %s, bindMount: %v: %v",
+			devicePath, podVolumeMapPath, volumeMapName, false, mapErr)
+	}
+
+	// Take file descriptor lock to keep a block device opened. Otherwise, there is a case
+	// that the block device is silently removed and attached another device with the same name.
+	// Container runtime can't handle this problem. To avoid unexpected condition fd lock
+	// for the block device is required.
+	_, mapErr = blkUtil.AttachFileDevice(filepath.Join(globalMapPath, string(podUID)))
 	if mapErr != nil {
-		return mapErr
+		return fmt.Errorf("blkUtil.AttachFileDevice failed. globalMapPath:%s, podUID: %s: %v",
+			globalMapPath, string(podUID), mapErr)
+	}
+
+	return nil
+}
+
+// UnmapBlockVolume is a utility function to provide a common way of unmapping
+// block device path for a specified volume and pod.  This function should be
+// called by volume plugins that implements volume.BlockVolumeMapper.Map() method.
+func UnmapBlockVolume(
+	blkUtil volumepathhandler.BlockVolumePathHandler,
+	globalUnmapPath,
+	podDeviceUnmapPath,
+	volumeMapName string,
+	podUID utypes.UID,
+) error {
+	// Release file descriptor lock.
+	err := blkUtil.DetachFileDevice(filepath.Join(globalUnmapPath, string(podUID)))
+	if err != nil {
+		return fmt.Errorf("blkUtil.DetachFileDevice failed. globalUnmapPath:%s, podUID: %s: %v",
+			globalUnmapPath, string(podUID), err)
+	}
+
+	// Ensure that all file descriptor lock for the path is released.
+	lockPath, err := blkUtil.GetLoopDevice(filepath.Join(globalUnmapPath, string(podUID)))
+	if err != nil {
+		if err.Error() != volumepathhandler.ErrDeviceNotFound {
+			return fmt.Errorf("blkUtil.GetLoopDevice failed. globalUnmapPath:%s, podUID: %s: %v",
+				globalUnmapPath, string(podUID), err)
+		}
+	} else if len(lockPath) != 0 {
+		return fmt.Errorf("Descriptor lock for globalUnmapPath:%s, podUID: %s is still taken. Wait for another retry to release it",
```

Comment:
How could this be possible if we just did DetachFileDevice in the lines above? What will be different in the next retry that it may succeed?

Reply:
I think that `MapBlockVolume` could be called more than once, so there would be a case that multiple descriptor locks are taken to the same `lockPath`, just for one volume-pod mapping. (I haven't actually hit this case, though.)

If above assumption is true, we might also need to return error if bindmount still exist after `UnmapDevice` for `globalUnmapPath` below this code, because bindmount could be created multiple times (On the other hand, symlink can't be created multiple times, so we won't need the same check for `podDeviceUnmapPath`).

Reply:
If the path is the same, i.e. a loop device already exists whose backing file is equal to the path, AttachFileDevice should no-op: https://github.com/kubernetes/kubernetes/blob/a3ccea9d8743f2ff82e41b6c2af6dc2c41dc7b10/pkg/volume/util/volumepathhandler/volume_path_handler_linux.go#L34. So we don't need to worry about that case.

Similarly, if the path is the same and you try to bind mount twice, wouldn't it error in MapBlockVolume?
mount --bind a b # 1st time
mount --bind a b # 2nd time, error?

Reply:
Thank you for your comment.

For bind mount, I've added a code to check if the path has already been bind mounted in `mapBindMountDevice`. So, it would be as safe as loop back device. 
However, IMHO, we don't take a lock between check and creation, so there could be a very small chance for creation logic to be called multiple times. 

And in such a situation, both `losetup` and `mount --bind` allow creating them every time, as seen below:

<details>

```
[losetup]
# losetup -f --show /dev/loop0 
/dev/loop1
# losetup -f --show /dev/loop0 
/dev/loop2
# losetup 
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /tmp/file1
/dev/loop1         0      0         0  0 /dev/loop0
/dev/loop2         0      0         0  0 /dev/loop0

[bind mount]
# touch /tmp/bindmount
# mount --bind /dev/loop0 /tmp/bindmount 
# mount --bind /dev/loop0 /tmp/bindmount 
# mount | grep /tmp/bindmount
devtmpfs on /tmp/bindmount type devtmpfs (rw,nosuid,seclabel,size=485080k,nr_inodes=121270,mode=755)
devtmpfs on /tmp/bindmount type devtmpfs (rw,nosuid,seclabel,size=485080k,nr_inodes=121270,mode=755)
```

</details>

Reply:
Thanks for testing. operation_generator/executor should insulate us from multiple creations  happening concurrently but it doesn't hurt to have these checks.

Reply:
Good to hear that operation_generator/executor avoid such a concurrent call. Then, I agree to delete this unnecessary code.